Client connection timeout when messaginguser updated concurrently

**Describe the bug**
If a messaginguser is created or modified at the same time as the client is connecting, an exception is thrown by the authentication handler, and the client connection times out.

**To Reproduce**
Steps to reproduce the behavior:
1. Run authorization tests on a CI environment
2. If test failures, look in keycloak log

**Expected behavior**
The authentication plugin should retry the operation N times before failing to reduce the likelihood of conflict. If retries are exceeded, fail the authentication.

**Additional context**


```
[0m2019-12-11T00:42:31.186Z WARN  [arjuna] ARJUNA012125: TwoPhaseCoordinator.beforeCompletion - failed for SynchronizationImple< 0:ffffac11000e:-7a7c2edf:5df01a06:c78c, org.wildfly.transaction.client.AbstractTransaction$AssociatingSynchronization@1238c355 >: javax.persistence.OptimisticLockException: Batch update returned unexpected row count from update [0]; actual row count: 0; expected: 1
	at org.hibernate.jpa.spi.AbstractEntityManagerImpl.wrapStaleStateException(AbstractEntityManagerImpl.java:1729)
	at org.hibernate.jpa.spi.AbstractEntityManagerImpl.convert(AbstractEntityManagerImpl.java:1634)
	at org.hibernate.jpa.spi.AbstractEntityManagerImpl.convert(AbstractEntityManagerImpl.java:1602)
	at org.hibernate.jpa.spi.AbstractEntityManagerImpl.convert(AbstractEntityManagerImpl.java:1608)
	at org.hibernate.jpa.internal.EntityManagerImpl$CallbackExceptionMapperImpl.mapManagedFlushFailure(EntityManagerImpl.java:235)
	at org.hibernate.internal.SessionImpl.flushBeforeTransactionCompletion(SessionImpl.java:3163)
	at org.hibernate.internal.SessionImpl.beforeTransactionCompletion(SessionImpl.java:2352)
	at org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.beforeTransactionCompletion(JdbcCoordinatorImpl.java:491)
	at org.hibernate.resource.transaction.backend.jta.internal.JtaTransactionCoordinatorImpl.beforeCompletion(JtaTransactionCoordinatorImpl.java:316)
	at org.hibernate.resource.transaction.backend.jta.internal.synchronization.SynchronizationCallbackCoordinatorNonTrackingImpl.beforeCompletion(SynchronizationCallbackCoordinatorNonTrackingImpl.java:47)
	at org.hibernate.resource.transaction.backend.jta.internal.synchronization.RegisteredSynchronization.beforeCompletion(RegisteredSynchronization.java:37)
	at org.wildfly.transaction.client.AbstractTransaction.performConsumer(AbstractTransaction.java:236)
	at org.wildfly.transaction.client.AbstractTransaction.performConsumer(AbstractTransaction.java:247)
	at org.wildfly.transaction.client.AbstractTransaction$AssociatingSynchronization.beforeCompletion(AbstractTransaction.java:292)
	at com.arjuna.ats.internal.jta.resources.arjunacore.SynchronizationImple.beforeCompletion(SynchronizationImple.java:76)
	at com.arjuna.ats.arjuna.coordinator.TwoPhaseCoordinator.beforeCompletion(TwoPhaseCoordinator.java:368)
	at com.arjuna.ats.arjuna.coordinator.TwoPhaseCoordinator.end(TwoPhaseCoordinator.java:91)
	at com.arjuna.ats.arjuna.AtomicAction.commit(AtomicAction.java:162)
	at com.arjuna.ats.internal.jta.transaction.arjunacore.TransactionImple.commitAndDisassociate(TransactionImple.java:1289)
	at com.arjuna.ats.internal.jta.transaction.arjunacore.BaseTransaction.commit(BaseTransaction.java:126)
	at com.arjuna.ats.jbossatx.BaseTransactionManagerDelegate.commit(BaseTransactionManagerDelegate.java:89)
	at org.wildfly.transaction.client.LocalTransaction.commitAndDissociate(LocalTransaction.java:77)
	at org.wildfly.transaction.client.ContextTransactionManager.commit(ContextTransactionManager.java:71)
	at org.keycloak.transaction.JtaTransactionWrapper.commit(JtaTransactionWrapper.java:92)
	at org.keycloak.services.DefaultKeycloakTransactionManager.commit(DefaultKeycloakTransactionManager.java:136)
	at io.enmasse.keycloak.spi.PlainSaslServerMechanism$1.processResponse(PlainSaslServerMechanism.java:144)
	at io.enmasse.keycloak.spi.SaslAuthenticator.process(SaslAuthenticator.java:171)
	at io.vertx.proton.impl.ProtonServerImpl$1.process(ProtonServerImpl.java:180)
	at io.vertx.proton.impl.ProtonTransport.processSaslAuthentication(ProtonTransport.java:211)
	at io.vertx.proton.impl.ProtonTransport.handleSocketBuffer(ProtonTransport.java:194)
	at io.vertx.core.net.impl.NetSocketImpl$DataMessageHandler.handle(NetSocketImpl.java:386)
	at io.vertx.core.net.impl.NetSocketImpl.lambda$new$2(NetSocketImpl.java:101)
	at io.vertx.core.streams.impl.InboundBuffer.handleEvent(InboundBuffer.java:235)
	at io.vertx.core.streams.impl.InboundBuffer.write(InboundBuffer.java:125)
	at io.vertx.core.net.impl.NetSocketImpl.handleMessage(NetSocketImpl.java:364)
	at io.vertx.core.impl.ContextImpl.executeTask(ContextImpl.java:320)
	at io.vertx.core.impl.EventLoopContext.execute(EventLoopContext.java:43)
	at io.vertx.core.impl.ContextImpl.executeFromIO(ContextImpl.java:188)
	at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:173)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:345)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:337)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:86)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:345)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:337)
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1476)
	at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1225)
	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1272)
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:502)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:441)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:278)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:345)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:337)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1408)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:345)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:930)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:677)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:612)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:529)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:491)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:905)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.hibernate.StaleStateException: Batch update returned unexpected row count from update [0]; actual row count: 0; expected: 1
	at org.hibernate.jdbc.Expectations$BasicExpectation.checkBatched(Expectations.java:67)
	at org.hibernate.jdbc.Expectations$BasicExpectation.verifyOutcome(Expectations.java:54)
	at org.hibernate.engine.jdbc.batch.internal.NonBatchingBatch.addToBatch(NonBatchingBatch.java:46)
	at org.hibernate.persister.entity.AbstractEntityPersister.update(AbstractEntityPersister.java:3198)
	at org.hibernate.persister.entity.AbstractEntityPersister.updateOrInsert(AbstractEntityPersister.java:3077)
	at org.hibernate.persister.entity.AbstractEntityPersister.update(AbstractEntityPersister.java:3457)
	at org.hibernate.action.internal.EntityUpdateAction.execute(EntityUpdateAction.java:145)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:589)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:463)
	at org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:337)
	at org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:39)
	at org.hibernate.internal.SessionImpl.flush(SessionImpl.java:1295)
	at org.hibernate.internal.SessionImpl.managedFlush(SessionImpl.java:468)
	at org.hibernate.internal.SessionImpl.flushBeforeTransactionCompletion(SessionImpl.java:3159)
	... 61 more
```

--- a/keycloak-plugin/sasl-plugin/src/main/java/io/enmasse/keycloak/spi/PlainSaslServerMechanism.java
+++ b/keycloak-plugin/sasl-plugin/src/main/java/io/enmasse/keycloak/spi/PlainSaslServerMechanism.java
@@ -39,6 +39,7 @@ import static io.enmasse.keycloak.spi.K8sServiceAccountCredentialProvider.ENMASS
 public class PlainSaslServerMechanism implements SaslServerMechanism {
 
     private static final Logger LOG = Logger.getLogger(PlainSaslServerMechanism.class);
+    private static final int MAX_RETRIES = 10;
 
     @Override
     public String getName() {
@@ -69,8 +70,7 @@ public class PlainSaslServerMechanism implements SaslServerMechanism {
             private RuntimeException error;
 
             @Override
-            public byte[] processResponse(byte[] response) throws IllegalArgumentException
-            {
+            public byte[] processResponse(byte[] response) throws IllegalArgumentException, InterruptedException {
                 if(error != null) {
                     throw error;
                 }
@@ -92,57 +92,74 @@ public class PlainSaslServerMechanism implements SaslServerMechanism {
                 String password = new String(response, authcidNullPosition + 1, passwordLen, StandardCharsets.UTF_8);
 
                 LOG.debug("SASL hostname: " + hostname);
-                KeycloakSession keycloakSession = keycloakSessionFactory.create();
-                KeycloakTransactionManager transactionManager = keycloakSession.getTransactionManager();
-                transactionManager.begin();
-                try {
-                    final RealmModel realm = keycloakSession.realms().getRealmByName(hostname);
-                    if (realm == null) {
-                        LOG.info("Realm " + hostname + " not found");
-                        authenticated = false;
-                        complete = true;
-                        return null;
-                    }
-                    if("@@serviceaccount@@".equals(username)) {
-                        String tokenUser = K8sServiceAccountCredentialProvider.authenticateToken(password, amqpServer.getOpenShiftClient(), amqpServer.getHttpClient());
-                        if(tokenUser != null) {
-                            final UserModel user = keycloakSession.userStorageManager().getUserByUsername(tokenUser, realm);
-                            if (user != null) {
-                                if ("serviceaccount".equals(user.getFirstAttribute("authenticationType"))) {
-                                    authenticatedUser = new UserDataImpl(user.getId(), user.getUsername(), user.getGroups().stream().map(GroupModel::getName).collect(Collectors.toSet()));
-                                    authenticated = true;
-                                    complete = true;
-                                    return null;
-                                } else {
-                                    LOG.debug("User " + tokenUser + " in realm " + hostname + " is not a serviceaccount user");
-                                }
-                            } else {
-                                LOG.debug("No such user " + tokenUser + " in realm " + hostname);
-                            }
-                        }
-                    } else {
-                        final UserModel user = keycloakSession.userStorageManager().getUserByUsername(username, realm);
-                        if (user != null) {
-                            UserCredentialModel credentialModel = "serviceaccount".equals(user.getFirstAttribute("authenticationType")) ? createServiceAccountUserCredential(password) : UserCredentialModel.password(password);
-                            if (keycloakSession.userCredentialManager().isValid(realm, user, credentialModel)) {
-                                authenticatedUser = new UserDataImpl(user.getId(), user.getUsername(), user.getGroups().stream().map(GroupModel::getName).collect(Collectors.toSet()));
-                                authenticated = true;
+                int retry = 0;
+                while (true) {
+                    KeycloakSession keycloakSession = keycloakSessionFactory.create();
+                    KeycloakTransactionManager transactionManager = keycloakSession.getTransactionManager();
+                    transactionManager.begin();
+                    try {
+                        try {
+                            final RealmModel realm = keycloakSession.realms().getRealmByName(hostname);
+                            if (realm == null) {
+                                LOG.info("Realm " + hostname + " not found");
+                                authenticated = false;
                                 complete = true;
                                 return null;
                             }
-                            LOG.debug("Invalid password for " + username + " in realm " + hostname);
-                        } else {
-                            LOG.debug("user not found: " + username);
+                            if ("@@serviceaccount@@".equals(username)) {
+                                String tokenUser = K8sServiceAccountCredentialProvider.authenticateToken(password, amqpServer.getOpenShiftClient(), amqpServer.getHttpClient());
+                                if (tokenUser != null) {
+                                    final UserModel user = keycloakSession.userStorageManager().getUserByUsername(tokenUser, realm);
+                                    if (user != null) {
+                                        if ("serviceaccount".equals(user.getFirstAttribute("authenticationType"))) {
+                                            authenticatedUser = new UserDataImpl(user.getId(), user.getUsername(), user.getGroups().stream().map(GroupModel::getName).collect(Collectors.toSet()));
+                                            authenticated = true;
+                                            complete = true;
+                                            return null;
+                                        } else {
+                                            LOG.debug("User " + tokenUser + " in realm " + hostname + " is not a serviceaccount user");
+                                        }
+                                    } else {
+                                        LOG.debug("No such user " + tokenUser + " in realm " + hostname);
+                                    }
+                                }
+                            } else {
+                                final UserModel user = keycloakSession.userStorageManager().getUserByUsername(username, realm);
+                                if (user != null) {
+                                    UserCredentialModel credentialModel = "serviceaccount".equals(user.getFirstAttribute("authenticationType")) ? createServiceAccountUserCredential(password) : UserCredentialModel.password(password);
+                                    if (keycloakSession.userCredentialManager().isValid(realm, user, credentialModel)) {
+                                        authenticatedUser = new UserDataImpl(user.getId(), user.getUsername(), user.getGroups().stream().map(GroupModel::getName).collect(Collectors.toSet()));
+                                        authenticated = true;
+                                        complete = true;
+                                        return null;
+                                    }
+                                    LOG.debug("Invalid password for " + username + " in realm " + hostname);
+                                } else {
+                                    LOG.debug("user not found: " + username);
+                                }
+
+                            }
+                            authenticated = false;
+                            complete = true;
+                            return null;
+
+                        } finally {
+                            transactionManager.commit();
+                            keycloakSession.close();
                         }
-
+                    } catch (Exception e) {
+                        if (retry < MAX_RETRIES) {
+                            LOG.warn("Exception when authenticating " + username, e);
+                            Thread.sleep(1000);
+                            retry++;
+                            LOG.info("Retry attempt " + retry + " for authenticating " + username);
+                        } else {
+                            LOG.warn("Exception when authenticating " + username + ". Aborting.", e);
+                            authenticated = false;
+                            complete = true;
+                            return null;
+                        }
                     }
-                    authenticated = false;
-                    complete = true;
-                    return null;
-
-                } finally {
-                    transactionManager.commit();
-                    keycloakSession.close();
                 }
             }
 

--- a/keycloak-plugin/sasl-plugin/src/main/java/io/enmasse/keycloak/spi/SaslAuthenticator.java
+++ b/keycloak-plugin/sasl-plugin/src/main/java/io/enmasse/keycloak/spi/SaslAuthenticator.java
@@ -187,9 +187,11 @@ class SaslAuthenticator implements ProtonSaslAuthenticator
                         sasl.done(Sasl.SaslOutcome.PN_SASL_AUTH);
                     }
                 }
-            }
-            catch (IllegalArgumentException e)
-            {
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                done = true;
+                sasl.done(Sasl.SaslOutcome.PN_SASL_SYS);
+            } catch (IllegalArgumentException e) {
                 e.printStackTrace();
                 done = true;
                 sasl.done(Sasl.SaslOutcome.PN_SASL_SYS);

--- a/keycloak-plugin/sasl-plugin/src/main/java/io/enmasse/keycloak/spi/SaslServerMechanism.java
+++ b/keycloak-plugin/sasl-plugin/src/main/java/io/enmasse/keycloak/spi/SaslServerMechanism.java
@@ -35,7 +35,7 @@ public interface SaslServerMechanism {
                          final Config.Scope config,
                          AmqpServer amqpServer);
     interface Instance {
-        byte[] processResponse(byte[] response) throws IllegalArgumentException;
+        byte[] processResponse(byte[] response) throws IllegalArgumentException, InterruptedException;
         boolean isComplete();
         boolean isAuthenticated();
         UserData getAuthenticatedUser();

--- a/keycloak-plugin/sasl-plugin/src/test/java/io/enmasse/keycloak/spi/PlainSaslServerMechanismTest.java
+++ b/keycloak-plugin/sasl-plugin/src/test/java/io/enmasse/keycloak/spi/PlainSaslServerMechanismTest.java
@@ -73,7 +73,7 @@ public class PlainSaslServerMechanismTest {
 
     // unknown realm
     @Test
-    public void testUnknownRealm() {
+    public void testUnknownRealm() throws InterruptedException {
         final SaslServerMechanism.Instance instance =
                 (new PlainSaslServerMechanism()).newInstance(keycloakSessionFactory, "unknownRealm", config, null);
         byte[] response = instance.processResponse(createInitialResponse("user", "password"));
@@ -85,7 +85,7 @@ public class PlainSaslServerMechanismTest {
 
     // known realm, unknown user
     @Test
-    public void testUnknownUser() {
+    public void testUnknownUser() throws InterruptedException {
         final SaslServerMechanism.Instance instance =
                 (new PlainSaslServerMechanism()).newInstance(keycloakSessionFactory, "realm", config, null);
         byte[] response = instance.processResponse(createInitialResponse("unknown", "password"));
@@ -97,7 +97,7 @@ public class PlainSaslServerMechanismTest {
 
     // known user, wrong password
     @Test
-    public void testWrongPassword() {
+    public void testWrongPassword() throws InterruptedException {
         final SaslServerMechanism.Instance instance =
                 (new PlainSaslServerMechanism()).newInstance(keycloakSessionFactory, "realm", config, null);
         byte[] response = instance.processResponse(createInitialResponse("user", "wrong"));
@@ -109,7 +109,7 @@ public class PlainSaslServerMechanismTest {
 
     // known user, correct password
     @Test
-    public void testCorrectPassword() {
+    public void testCorrectPassword() throws InterruptedException {
         final SaslServerMechanism.Instance instance =
                 (new PlainSaslServerMechanism()).newInstance(keycloakSessionFactory, "realm", config, null);
         byte[] response = instance.processResponse(createInitialResponse("user", "password"));

--- a/keycloak-plugin/sasl-plugin/src/test/java/io/enmasse/keycloak/spi/XOAUTH2SaslServerMechanismTest.java
+++ b/keycloak-plugin/sasl-plugin/src/test/java/io/enmasse/keycloak/spi/XOAUTH2SaslServerMechanismTest.java
@@ -70,7 +70,7 @@ public class XOAUTH2SaslServerMechanismTest implements XOAUTH2SaslServerMechanis
 
     // unknown realm
     @Test
-    public void testUnknownRealm() {
+    public void testUnknownRealm() throws InterruptedException {
         final SaslServerMechanism.Instance instance =
                 (new XOAUTH2SaslServerMechanism(this)).newInstance(keycloakSessionFactory, "unknownRealm", config, null);
         byte[] response = instance.processResponse(createInitialResponse("user", "token"));
@@ -83,7 +83,7 @@ public class XOAUTH2SaslServerMechanismTest implements XOAUTH2SaslServerMechanis
 
     // Invalid token
     @Test
-    public void testWrongPassword() {
+    public void testWrongPassword() throws InterruptedException {
         final SaslServerMechanism.Instance instance =
                 (new XOAUTH2SaslServerMechanism(this)).newInstance(keycloakSessionFactory, "realm", config, null);
         this.verifyToken = false;
@@ -95,7 +95,7 @@ public class XOAUTH2SaslServerMechanismTest implements XOAUTH2SaslServerMechanis
 
     // Validated token
     @Test
-    public void testValidatedToken() {
+    public void testValidatedToken() throws InterruptedException {
         final SaslServerMechanism.Instance instance =
                 (new XOAUTH2SaslServerMechanism(this)).newInstance(keycloakSessionFactory, "realm", config, null);
         this.verifyToken = true;

--- a/pkg/controller/messaginguser/messaginguser_controller.go
+++ b/pkg/controller/messaginguser/messaginguser_controller.go
@@ -104,7 +104,7 @@ func (r *ReconcileMessagingUser) Reconcile(request reconcile.Request) (reconcile
 
 	rc := &recon.ReconcileContext{}
 
-	reqLogger.V(8).Info("checkFinalizer", "user", user.ObjectMeta)
+	reqLogger.V(2).Info("checkFinalizer", "user", user.ObjectMeta)
 	rc.Process(func() (reconcile.Result, error) {
 		return r.checkFinalizer(ctx, reqLogger, user)
 	})
@@ -126,7 +126,7 @@ func (r *ReconcileMessagingUser) Reconcile(request reconcile.Request) (reconcile
 		return nil
 	})
 
-	reqLogger.V(8).Info("createOrUpdate", "user", user.ObjectMeta)
+	reqLogger.V(2).Info("createOrUpdate", "user", user.ObjectMeta)
 	rc.Process(func() (reconcile.Result, error) {
 		if user.Status.Phase != userv1beta1.UserTerminating {
 			return r.createOrUpdateUser(ctx, reqLogger, user)
@@ -138,7 +138,7 @@ func (r *ReconcileMessagingUser) Reconcile(request reconcile.Request) (reconcile
 		user.Status.Message = rc.Error().Error()
 	}
 
-	reqLogger.V(8).Info("updateGeneration", "user", user.ObjectMeta)
+	reqLogger.V(2).Info("updateGeneration", "user", user.ObjectMeta)
 	rc.ProcessSimple(func() error {
 		if user.Status.Message != lastMessage || user.Generation != user.Status.Generation {
 			user.Status.Generation = user.Generation

--- a/pkg/keycloak/client.go
+++ b/pkg/keycloak/client.go
@@ -128,6 +128,7 @@ func (c *keycloakClient) CreateUser(realm string, user *userv1beta1.MessagingUse
 	if err != nil {
 		return err
 	}
+	log.V(4).Info("Created user", "username", newUser.Username, "userId", userId)
 
 	err = c.applyAuthentication(realm, user.Spec.Authentication, userId)
 	if err != nil {
@@ -172,6 +173,7 @@ func (c *keycloakClient) applyAuthentication(realm string, authSpec userv1beta1.
 	switch authSpec.Type {
 	case userv1beta1.Password:
 		if authSpec.Password != nil {
+			log.V(4).Info("Setting password", "userId", userId)
 			err := c.client.SetPassword(c.token.AccessToken, userId, realm, string(authSpec.Password), false)
 			if err != nil {
 				return err
@@ -225,7 +227,7 @@ func (c *keycloakClient) applyAuthorization(realm string, authzSpec []userv1beta
 	}
 
 	if !reflect.DeepEqual(existingGroups, desiredGroups) {
-		log.Info("Changing user groups", "user", userId, "realm", realm, "old", existingGroups, "new", desiredGroups)
+		log.V(4).Info("Changing user groups", "userId", userId, "realm", realm, "old", existingGroups, "new", desiredGroups)
 
 		for groupName := range existingGroups {
 			if !desiredGroups[groupName] {

--- a/pkg/util/finalizer/finalizer.go
+++ b/pkg/util/finalizer/finalizer.go
@@ -18,7 +18,7 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
-var log = logf.Log.WithName("controller_iotproject")
+var log = logf.Log.WithName("finalizer")
 
 type DeconstructorContext struct {
 	Context context.Context

